src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java

```diff
@@ -129,19 +143,65 @@ public void run(ActionContext context) throws Exception {
 
     QueryJobConfiguration queryConfig = builder.build();
 
-    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+    // Exponential backoff with initial retry of 1 second and max retry of 32 seconds.
+    executeQueryWithExponentialBackoff(bigQuery, queryConfig, context);
+  }
+
+  protected void executeQueryWithExponentialBackoff(BigQuery bigQuery,
+                                                   QueryJobConfiguration queryConfig, ActionContext context)
+          throws BigQueryJobExecutionException {
+    try {
+      Failsafe.with(getRetryPolicy()).run(() -> executeQuery(bigQuery, queryConfig, context));
+    } catch (RuntimeException e) {
+      LOG.error("Failed to execute BigQuery job. Error: {}", e.getMessage());
+      throw new BigQueryJobExecutionException("Failed to execute BigQuery job. Reason: " + e.getMessage());
```

Comment:
This way you are loosing all the additional information from original BigQueryException stack trace. There are only some case where you would want to hide all details of the underlying exception by using e.getMessage instead of attaching a root cause. I don't see why would we want to have it here. Same applies to the catch clause in `executeQuery`

Reply:
Updated !